README.md
# rhfa-react-native

This library allows your React application to automatically generate forms using [ReactHookForm](https://react-hook-form.com/) that are redered by [React Native](https://reactnative.dev/). The form and validations are generated following a schema inspired by [SimpleSchema](https://github.com/aldeed/simple-schema-js).

It uses some react-native-community packages to help with checkbox, slider and picker.

## Upgrade to `1.2.0`

You must change your `package.json` entry from `@react-native-community/picker` to `@react-native-picker/picker`. It will look like this:

```
    "dependencies": {
      "@react-native-picker/picker": "^2.0.0",
      ...
    }
```

## Installation

    $ npm install react-hook-form rhfa-react-native @react-native-community/checkbox @react-native-community/slider @react-native-picker/picker --save

## Usage

```javascript
    import { createSchema, Autoform } from 'rhfa-react-native'
    import { createStyles } from 'rhfa-react-native/dist/emergencyStyles'

    const style = createStyles()

    export const client = createSchema('client', {
      name: {
        type: 'string',
        required: true,
        max: 32
      },
      age: {
        type: 'number'
      }
    })

    const MyForm = ({ onSubmit }) =>
      <Autoform
        schema={client}
        onSubmit={onSubmit}
        styles={styles}
      />
```

## Styling

The emergency styles are separated in its own bundle. The function `createStyles` will generate some styles for the components. It gereates some styles proceduraly and you can set some better defaults:

```javascript
  createStyles({
    backgroundColor = '#e0e0e0',
    height = 32,
    headerColor = '#a0a0a0',
    panelBorderColor = '#a0a0a0',
    inputBorderColor = '#c0c0c0',
    inputBackgroundColor = '#ffffff',
    radioRadio = 8,
    radioColor = '#808080',
    radioSelectedColor = '#404040',
    buttonColor = '#909090',
    addColor = '#70a070',
    addTextColor = '#104010',
    removeColor = '#a07070',
    removeTextColor = '#ffe0e0',
    errorColor = '#601010',
    errorBackgroundColor = '#ffe0e0',
    miniHeight = 24,
    fontSize = 16
  } = {})

```

## Specific schema attributes for React Native

### type = 'range'

Appart from `min` and `max`, you should setup `step`. It will coerce to nearest step:

```javascript
  something: {
    type: 'range',
    min: 10,
    max: 90,
    step: 5,
    defaultValue: 50
  }
```

### Helper text

You can specify helperText in the schema.

```javascript
    import { createSchema } from 'rhfa-react-native'

    const smt = createSchema('something', {
      name: {
        type: 'string',
        helperText: tr('models.name.helper')
      }
    })
```

You can set the text directly too, without using `tr()`.

Also you can set model translation string attribute `_helper` to any value:

```javascript
    setTranslations({
      name: {
        _field: 'Name',
        _helper: 'Please, enter your name'
      }
    })
```

### Label info

Labels can have a field that appears in a less prominent color adding some clarification. This label info will come from translations:

```javascript
    setTranslations({
      name: {
        _field: 'Name',
        _labelInfo: '(given name)'
      }
    })
```

The label `tr('requiredLabel')` will be added automatically for required fields (unless you override with translations).

## [Documentation](https://github.com/dgonz64/react-hook-form-auto)


CHANGELOG.md
# Changelog

## 1.2.0

* Correct helperText management
* __Breaking change__: switched from `@react-native-community/picker` to `@react-native-picker/picker`

## 1.1.0

* Updated to react-hook-form-auto 1.3

## 1.0.2

* Corrected README creating styles

## 1.0.1

### Fixed

* Correcting this file
* Doc

## 1.0.0

Initial!


test/submit.react.test.js
import React from 'react'
import { mount } from 'enzyme'
import { act } from 'react-dom/test-utils'

import config from './enzymeConfig'
import {
  createSchema,
  Autoform,
} from '../src/index'

const custom = createSchema('simple', {
  name: {
    type: 'string',
  },
  numberer: {
    type: 'number'
  },
  radiator: {
    type: 'radios',
    options: ['a', 'b']
  },
  booler: {
    type: 'boolean'
  }
})

test('Submits', async () => {
  let doSubmit
  const wasSubmitted = new Promise((resolve) => {
    doSubmit = resolve 
  })
  const mockSubmit = jest.fn(() => {
    doSubmit()
  })

  const app = mount(
    <Autoform schema={custom} onSubmit={mockSubmit} />
  )

  const input = app.find('input[name="name"]')
  input.instance().value = 'Hello'

  const numberer = app.find('input[name="numberer"]')
  numberer.instance().value = 128

  const radiator = app.find('input[name="radiator"][value="b"]')
  radiator.instance().checked = true

  const booler = app.find('input[name="booler"]')
  booler.instance().checked = true

  const Form = app.find(Autoform)
  await act(async () => {
    await Form.simulate('submit')
  })

  expect.assertions(2)
  const { calls } = mockSubmit.mock
  return wasSubmitted.then(() => {
    expect(calls.length).toBe(1)
    expect(calls[0][0]).toStrictEqual({
      name: 'Hello',
      numberer: 128,
      radiator: 'b',
      booler: true
    })
  })
})


src/skin.js
import React, { useEffect } from 'react'
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
} from 'react-native'
import { ArrayTable } from './components/ArrayTable'
import { ArrayPanel } from './components/ArrayPanel'
import { Radios } from './components/Radios'
import { Panel } from './components/Panel'
import { Button } from './components/Button'
import { TextInput } from './components/TextInput'
import {
  trPath,
  trField,
  tr,
  processOptions,
  stringExists
} from 'react-hook-form-auto'
import Checkbox from '@react-native-community/checkbox'
import Slider from '@react-native-community/slider'
import { Picker } from '@react-native-picker/picker'

function createChanger(setValue, name) {
  return function onChange(value) {
    setValue(name, value, { shouldValidate: true })
  }
}

const GroupAdaptor = ({
  name,
  field,
  fieldSchema,
  schemaTypeName,
  errorText,
  helperText,
  inline,
  children,
  labelOverride,
  addWrapperProps,
  styles
}) => {
  if (inline) {
    return children
  } else {
    const label = typeof labelOverride != 'undefined' ?
      labelOverride : trField({ fieldSchema, schemaTypeName, field })

    if (errorText)
      helperText = errorText

    const infoId = trPath(schemaTypeName, field, '_labelInfo')
    const labelInfo = stringExists(infoId) ?
      tr(infoId) : fieldSchema.required && tr('requiredLabel')

    let containerStyles = [
      styles.container,
      errorText && styles.erroredContainer
    ].filter(style => style)

    let helperTextStyles = [
      styles.helperText,
      errorText && styles.erroredHelperText
    ]

    return (
      <View style={containerStyles}>
        <View style={styles.labelContainer}>
          <Text style={styles.label}>
            {label}
          </Text>  
          {
            labelInfo &&
              <Text style={styles.labelInfo}>
                {labelInfo}
              </Text>
          }
        </View>
        {children}
        {
          helperText &&
            <Text style={helperTextStyles}>
              {helperText}
            </Text>
        }
      </View>
    )
  }
}

// Both render and adaptorComponent will do.
const ControlAdaptor = ({
  name,
  defaultValue,
  value,
  errorText,
  field,
  fieldSchema,
  formHook,
  rules,
  adaptorComponent,
  children,
  controlProps,
  setValue,
  styles,
  render,
  ...rest
}) => {
  let inputStyle = [
    styles.input,
    errorText && styles.erroredInput
  ]

  // Will use my own onChange instead of Control's to
  // make it work with coercers
  const handleChange = createChanger(setValue, name)

  const Comp = adaptorComponent

  return (
    <Comp
      {...controlProps}
      {...rest}
      value={value}
      onChange={handleChange}
      styles={styles}
      style={inputStyle}
    />
  )
}

export default {
  defaultWrap: GroupAdaptor,
  string: {
    controlled: true,
    render: (props) => {
      return {
        ...props,
        component: ControlAdaptor,
        adaptorComponent: TextInput
      }
    }
  },
  number: {
    coerce: value => value && parseFloat(value) || 0,
    controlled: true,
    render: (props) => {
      return {
        ...props,
        component: ControlAdaptor,
        adaptorComponent: TextInput,
        controlProps: { type: 'number' }
      }
    }
  },
  password: {
    controlled: true,
    render: (props) => {
      return {
        ...props,
        component: ControlAdaptor,
        adaptorComponent: TextInput,
        controlProps: { type: 'password' }
      }
    }
  },
  select: {
    controlled: true,
    render: {
      component: (props) => {
        const {
          schemaTypeName,
          name,
          formHook,
          defaultValue,
          styles,

          value,
          setValue,
        } = props

        const label = trField(props)
        const options = processOptions({
          ...props,
          addDefault: true
        })

        const handleChange = createChanger(setValue, name)

        return (
          <Picker
            selectedValue={value}
            onValueChange={handleChange}
            style={styles.select}
          >
            {
              options.map(({ label, value }) =>
                <Picker.Item
                  key={value}
                  label={label}
                  value={value}
                />
              )
            }
          </Picker>
        )
      }
    }
  },
  boolean: {
    wrapper: (props) => props.children,
    coerce: value => Boolean(value),
    controlled: true,
    render: {
      component: (props) => {
        const {
          name,
          defaultValue,
          formHook,
          control,
          styles,

          value,
          setValue,
          onBlur
        } = props

        const label = trField(props)

        const handleChange = createChanger(setValue, name)

        return (
          <View style={styles.checkboxWrap}>
            <Checkbox
              key={name}
              name={name}
              value={value}
              onValueChange={handleChange}
            />
            <Text style={styles.checkboxLabel}>
              {label}
            </Text>
          </View>
        )
      }
    }
  },
  radios: {
    controlled: true,
    render: {
      component: (props) => {
        const {
          name,
          formHook,
          defaultValue,
          styles,

          value,
          setValue,
          onBlur
        } = props

        const label = trField(props)
        const options = processOptions(props)

        const handleChange = createChanger(setValue, name)

        return (
          <Radios
            options={options}
            value={value}
            onChange={handleChange}
            styles={styles}
          />
        )
      }
    }
  },
  range: {
    coerce: value => parseFloat(value),
    controlled: true,
    render: {
      component: (props) => {
        const {
          name,
          fieldSchema: {
            min,
            max,
            step,
            sliderParams
          },
          formHook,
          styles,

          value,
          setValue,
          onBlur
        } = props

        const defaultValue = typeof props.defaultValue == 'undefined' ?
          min : props.defaultValue
        const stepExp = Math.log10(step)
        const stepDecimals = stepExp >= 0 ? 0 : -stepExp

        const change = createChanger(setValue, name)

        const handleChange = (newValue) => {
          if (!step) {
            change(newValue)
          } else {
            const steps = parseInt((newValue - min) / step)
            const steppedValue = min + steps * step
            const prettyValue = steppedValue.toFixed(stepDecimals)
            change(parseFloat(prettyValue))
          }
        }

        return (
          <View style={styles.sliderContainer}>
            <Text style={styles.sliderValue}>{value}</Text>
            <Slider
              {...sliderParams}
              minimumValue={min}
              maximumValue={max}
              value={value}
              step={step}
              onValueChange={handleChange}
              style={styles.slider}
            />
          </View>
        )
      }
    }
  },
  button: {
    render: Button
  },
  arrayButton: {
    render: (props) => <Button {...props} mini />
  },
  form: {
    render: ({ children }) =>
      <View>
        {children}
      </View>
  },
  panel: {
    render: ({ children, header, styles }) =>
      <Panel header={header} styles={styles}>
        {children}
      </Panel>
  },
  addGlyph: {
    render: ({ styles }) =>
      <Text style={styles && styles.addText}>+</Text>
  },
  removeGlyph: {
    render: ({ styles }) =>
      <Text style={styles && styles.removeText}>-</Text>
  },
  arrayTable: {
    render: ArrayTable
  },
  arrayPanel: {
    render: ArrayPanel
  },
  div: {
    render: ({ className, ...rest }) =>
      <View style={className} {...rest} />
  },
  text: {
    render: ({ className, ...rest }) =>
      <Text style={className} {...rest} />
  }
}


webpack.config.js
const path = require('path')
const package = require('./package.json')

const { NODE_ENV } = process.env

module.exports = {
  mode: NODE_ENV || 'development',
  devtool: 'source-map',
  entry: {
    'rhfa-react-native': './src/index.js',
    'emergencyStyles': './src/emergencyStyles.js'
  },
  output: {
    path: path.join(__dirname, 'dist'),
    filename: `[name]${NODE_ENV === 'production' ? '.min' : ''}.js`,
    libraryTarget: 'umd',
  },
  resolve: {
    extensions: ['.js', '.jsx'],
    alias: {
      'react-hook-form-auto': 'react-hook-form-auto/dist/base.js'
    }
  },
  module: {
    rules: [
      {
        test: /\.(js|jsx)$/,
        exclude: /^node_modules/,
        loader: 'babel-loader',
      },
    ],
  },
  externals: {
    'react': 'react',
    'react-native': 'react-native',
    'react-hook-form': 'react-hook-form',
    '@react-native-community/checkbox': '@react-native-community/checkbox',
    '@react-native-community/slider': '@react-native-community/slider',
    '@react-native-picker/picker': '@react-native-picker/picker'
  }
}


src/components/Button.js
import React from 'react'
import {
  Button as RNButton,
  TouchableOpacity
} from 'react-native'

export const Button = ({
  className,
  styles,
  children,
  type,
  onClick,
  intent,
  mini,
  ...rest
}) => {
  styles = styles || className
  const rnIntent = type == 'submit' ? 'primary' : null

  let buttonStyle = [
    styles.button,
    mini && styles.buttonMini,
    intent == 'add' && styles.buttonAdd,
    intent == 'remove' && styles.buttonRemove,
  ].filter(style => style)

  if (typeof children == 'string') {
    return (
      <RNButton
        title={children}
        intent={rnIntent}
        style={buttonStyle}
        onPress={onClick}
        {...rest}
      />
    )
  } else {
    return (
      <TouchableOpacity
        style={buttonStyle}
        onPress={onClick}
        {...rest}
      >
        {children}
      </TouchableOpacity>
    )
  }
}


src/components/Radios.js
import React from 'react'
import {
  View,
  Text,
  TouchableWithoutFeedback
} from 'react-native'

export const Radios = ({
  options,
  value,
  onChange,
  styles
}) =>
  <View style={styles.radioContainer}>
    {
      options.map(option => {
        const handleChange = () => {
          onChange(option.value)
        }

        return (
          <TouchableWithoutFeedback
            key={option.value}
            onPress={handleChange}
          >
            <View style={styles.radioLine}>
              <View style={styles.radioWrap}>
                {
                  option.value == value &&
                    <View style={styles.radioSelected} />
                }
              </View>
              <Text style={styles.radioLabel}>
                {option.label}
              </Text>
            </View>
          </TouchableWithoutFeedback>
        )
      })
    }
  </View>


src/components/ArrayTable.js
import React from 'react'
// import { renderLectures } from './renderLectures'

const { Column, Table, Cell } = {}

import { trField } from 'react-hook-form-auto'

const renderColumns = ({ items, subType, fieldNames }) => {
  const someItem = items[0]

  const createCellRenderer = (column, colIdx) => rowIdx => {
    return (
      <Cell key={column}>
        {items[rowIdx].inputs[colIdx]}
      </Cell>
    )
  }

  return fieldNames.map((columnName, colIdx) => {
    const displayName = trField(subType, columnName)

    return (
      <Column
        key={columnName}
        name={displayName}
        cellRenderer={createCellRenderer(columnName, colIdx)}
      />
    )
  })
}

export const ArrayTable = ({ items, schema }) => {
  const subType = schema.getType()
  const schemaDef = schema.getSchema()
  const fieldNames = Object.keys(schemaDef)

  if (items.length > 0) {
    const renderRemove = rowIdx => {
      const { idx, closeButton } = items[rowIdx]

      return (
        <Cell key="remove">
          {closeButton}
        </Cell>
      )
    }

    console.log("ITEMS", items)

    return (
      <>
        {/* renderLectures(props) */}
        <Table numRows={items.length} defaultRowHeight={31}>
          {renderColumns({ items, subType, fieldNames })}
          <Column
            key="remove"
            name=""
            cellRenderer={renderRemove}
          />
        </Table>
      </>
    )
  } else
    return null
}


src/emergencyStyles.js
import { StyleSheet } from 'react-native'

export function createStyles({
  backgroundColor = '#e0e0e0',
  height = 32,
  headerColor = '#a0a0a0',
  panelBorderColor = '#a0a0a0',
  inputBorderColor = '#c0c0c0',
  inputBackgroundColor = '#ffffff',
  radioRadio = 8,
  radioColor = '#808080',
  radioSelectedColor = '#404040',
  buttonColor = '#909090',
  addColor = '#70a070',
  addTextColor = '#104010',
  removeColor = '#a07070',
  removeTextColor = '#ffe0e0',
  errorColor = '#601010',
  errorBackgroundColor = '#ffe0e0',
  miniHeight = 24,
  fontSize = 16
} = {}) {
  return StyleSheet.create({
    button: {
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: buttonColor,
      width: height,
      height
    },
    buttonMini: {
      width: miniHeight,
      height: miniHeight,
    },
    panel: {
      borderColor: panelBorderColor,
      marginBottom: 2,
      borderWidth: 1,
      borderRadius: 4,
      borderStyle: 'solid',
    },
    panelHeader: {
      flex: 1,
      flexDirection: 'row',
      paddingVertical: 12,
      paddingHorizontal: 4,
      backgroundColor,
      shadowColor: '#101010',
      shadowOffset: {
        width: 0,
        height: 1
      },
      shadowOpacity: 0.8,
      shadowRadius: 2,
      elevation: 3
    },
    panelHeaderText: {
      fontSize: fontSize - 2,
      fontWeight: 'bold'
    },
    panelContents: {
      paddingVertical: 4,
      paddingHorizontal: 8
    },
    container: {
      marginVertical: 4,
      paddingHorizontal: 4
    },
    erroredContainer: {
      backgroundColor: errorBackgroundColor,
      borderRadius: 2
    },
    input: {
      borderWidth: 1,
      borderStyle: 'solid',
      borderColor: inputBorderColor,
      backgroundColor: inputBackgroundColor,
      borderRadius: 4,
      paddingVertical: 8,
      fontSize,
      height
    },
    erroredInput: {
      borderWidth: 1.5,
      borderStyle: 'solid',
      borderColor: errorColor
    },
    labelContainer: {
      flex: 1,
      flexDirection: 'row',
      alignItems: 'center'
    },
    label: {
      fontWeight: 'bold',
      marginVertical: 2
    },
    labelInfo: {
      color: '#404040',
      fontSize: fontSize - 2,
      marginLeft: 4
    },
    helperText: {
      fontSize: fontSize - 4,
      color: '#606060'
    },
    erroredHelperText: {
      fontStyle: 'italic',
      color: errorColor
    },
    error: {
      fontSize: fontSize - 4,
      fontStyle: 'italic',
      color: errorColor
    },
    radioContainer: {
      flex: 1,
      flexDirection: 'column'
    },
    radioLine: {
      flex: 1,
      flexDirection: 'row',
      marginVertical: 2,
      alignItems: 'center'
    },
    radioWrap: {
      width: 2 * radioRadio,
      height: 2 * radioRadio,
      borderRadius: radioRadio,
      borderWidth: 2,
      borderColor: radioColor,
      alignItems: 'center',
      justifyContent: 'center',
      marginHorizontal: 6
    },
    radioSelected: {
      width: radioRadio,
      height: radioRadio,
      borderRadius: radioRadio / 2,
      backgroundColor: radioSelectedColor
    },
    radioLabel: {
      fontSize: fontSize - 2,
      marginLeft: 6
    },
    checkboxWrap: {
      flex: 1,
      flexDirection: 'row',
      alignItems: 'center',
      marginLeft: 2
    },
    checkboxLabel: {
      fontSize: fontSize - 2,
      marginLeft: 4
    },
    sliderContainer: {
      flex: 1,
      flexDirection: 'column',
      alignItems: 'stretch'
    },
    sliderValue: {
      alignSelf: 'center'
    },
    buttonAdd: {
      backgroundColor: addColor,
      borderRadius: 2,
      shadowColor: '#102010',
      shadowOpacity: 0.8,
      shadowRadius: 2,
      elevation: 2
    },
    buttonRemove: {
      backgroundColor: removeColor,
      borderRadius: 2,
      shadowColor: '#102010',
      shadowOpacity: 0.8,
      shadowRadius: 2,
      elevation: 2
    },
    addText: {
      color: addTextColor
    },
    removeText: {
      color: removeTextColor
    },
    inputPanelWrap: {
      flex: 1,
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center'
    },
    inputPanelEntity: {
      fontSize: fontSize,
      fontWeight: 'bold'
    }
  })
}
